File: sdk/swift/Sources/OperonCoreDriver/Memory.swift
import Foundation

public enum OperonMemoryKind: String, Codable, Sendable {
  case fact, preference, decision, episode
}

public enum OperonMemoryAuthority: String, Codable, Sendable {
  case applicationVerified = "application_verified"
  case userConfirmed = "user_confirmed"
  case userStated = "user_stated"
  case modelInferred = "model_inferred"
  case importedUntrusted = "imported_untrusted"
}

public enum OperonMemorySensitivity: String, Codable, Sendable {
  case `private`, `internal`, `public`
}

public enum OperonMemoryStatus: String, Codable, Sendable {
  case active, superseded, tombstoned
}

public struct OperonMemoryScope: Codable, Sendable, Equatable {
  public let namespace: String
  public let subject: String?
  public let allowedSensitivities: [OperonMemorySensitivity]

  public init(
    namespace: String,
    subject: String? = nil,
    allowedSensitivities: [OperonMemorySensitivity] = [.private, .internal]
  ) {
    precondition(!namespace.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
    precondition(!allowedSensitivities.isEmpty)
    self.namespace = namespace
    self.subject = subject
    self.allowedSensitivities = allowedSensitivities
  }

  enum CodingKeys: String, CodingKey {
    case namespace, subject
    case allowedSensitivities = "allowed_sensitivities"
  }
}

public struct OperonMemoryRecord: Codable, Sendable, Equatable {
  public let id: String
  public let namespace: String
  public let subject: String?
  public let kind: OperonMemoryKind
  public let content: String
  public let authority: OperonMemoryAuthority
  public let sensitivity: OperonMemorySensitivity
  public let confidence: Double?
  public let sourceIDs: [String]
  public let occurredAt: String?
  public let observedAt: String
  public let validFrom: String?
  public let validUntil: String?
  public let supersedes: String?
  public let status: OperonMemoryStatus
  public let createdBy: String
  public let schemaVersion: Int

  public init(
    id: String = UUID().uuidString.lowercased(),
    namespace: String,
    subject: String? = nil,
    kind: OperonMemoryKind,
    content: String,
    authority: OperonMemoryAuthority,
    sensitivity: OperonMemorySensitivity = .private,
    confidence: Double? = nil,
    sourceIDs: [String] = [],
    occurredAt: String? = nil,
    observedAt: String = ISO8601DateFormatter().string(from: Date()),
    validFrom: String? = nil,
    validUntil: String? = nil,
    supersedes: String? = nil,
    status: OperonMemoryStatus = .active,
    createdBy: String = "application",
    schemaVersion: Int = 1
  ) {
    precondition(!namespace.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
    precondition(!content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
    if let confidence { precondition((0...1).contains(confidence)) }
    self.id = id
    self.namespace = namespace
    self.subject = subject
    self.kind = kind
    self.content = content
    self.authority = authority
    self.sensitivity = sensitivity
    self.confidence = confidence
    self.sourceIDs = sourceIDs
    self.occurredAt = occurredAt
    self.observedAt = observedAt
    self.validFrom = validFrom
    self.validUntil = validUntil
    self.supersedes = supersedes
    self.status = status
    self.createdBy = createdBy
    self.schemaVersion = schemaVersion
  }

  enum CodingKeys: String, CodingKey {
    case id, namespace, subject, kind, content, authority, sensitivity, confidence, status
    case sourceIDs = "source_ids"
    case occurredAt = "occurred_at"
    case observedAt = "observed_at"
    case validFrom = "valid_from"
    case validUntil = "valid_until"
    case supersedes = "supersedes"
    case createdBy = "created_by"
    case schemaVersion = "schema_version"
  }
}

public protocol OperonMemoryStore: Sendable {
  func search(
    _ query: String,
    scope: OperonMemoryScope,
    limit: Int
  ) async throws -> [OperonMemoryRecord]
}

/// Small, durable, application-owned memory store for Apple hosts.
///
/// It writes one JSON document atomically. The protocol boundary makes it
/// replaceable with a SQLite implementation without changing the core driver.
public actor FileOperonMemoryStore: OperonMemoryStore {
  private let url: URL
  private var records: [OperonMemoryRecord]

  public init(url: URL) throws {
    self.url = url
    if FileManager.default.fileExists(atPath: url.path) {
      records = try JSONDecoder().decode([OperonMemoryRecord].self, from: Data(contentsOf: url))
    } else {
      records = []
    }
  }

  @discardableResult
  public func put(_ record: OperonMemoryRecord) throws -> OperonMemoryRecord {
    if let supersedes = record.supersedes,
      let index = records.firstIndex(where: {
        $0.id == supersedes && $0.namespace == record.namespace
      })
    {
      records[index] = withStatus(records[index], .superseded)
    }
    records.removeAll { $0.id == record.id }
    records.append(record)
    try persist()
    return record
  }

  public func search(
    _ query: String,
    scope: OperonMemoryScope,
    limit: Int
  ) async throws -> [OperonMemoryRecord] {
    precondition(limit > 0)
    let now = ISO8601DateFormatter().string(from: Date())
    let terms = query.lowercased().split { !$0.isLetter && !$0.isNumber }.map(String.init)
    return
      records
      .filter { record in
        record.namespace == scope.namespace
          && (scope.subject == nil || record.subject == scope.subject)
          && scope.allowedSensitivities.contains(record.sensitivity)
          && record.status == .active
          && (record.validFrom == nil || record.validFrom! <= now)
          && (record.validUntil == nil || record.validUntil! > now)
      }
      .sorted {
        score($0, terms: terms) == score($1, terms: terms)
          ? $0.observedAt > $1.observedAt
          : score($0, terms: terms) > score($1, terms: terms)
      }
      .prefix(limit)
      .map { $0 }
  }

  public func tombstone(_ id: String) throws -> Bool {
    guard let index = records.firstIndex(where: { $0.id == id }) else { return false }
    records[index] = withStatus(records[index], .tombstoned)
    try persist()
    return true
  }

  public func export(scope: OperonMemoryScope) -> [OperonMemoryRecord] {
    records.filter {
      $0.namespace == scope.namespace && (scope.subject == nil || $0.subject == scope.subject)
    }
  }

  public func delete(namespace: String) throws -> Int {
    let before = records.count
    records.removeAll { $0.namespace == namespace }
    try persist()
    return before - records.count
  }

  private func persist() throws {
    try FileManager.default.createDirectory(
      at: url.deletingLastPathComponent(),
      withIntermediateDirectories: true
    )
    let data = try JSONEncoder().encode(records)
    try data.write(to: url, options: .atomic)
  }
}

private func withStatus(_ record: OperonMemoryRecord, _ status: OperonMemoryStatus)
  -> OperonMemoryRecord
{
  OperonMemoryRecord(
    id: record.id, namespace: record.namespace, subject: record.subject, kind: record.kind,
    content: record.content, authority: record.authority, sensitivity: record.sensitivity,
    confidence: record.confidence, sourceIDs: record.sourceIDs, occurredAt: record.occurredAt,
    observedAt: record.observedAt, validFrom: record.validFrom, validUntil: record.validUntil,
    supersedes: record.supersedes, status: status, createdBy: record.createdBy,
    schemaVersion: record.schemaVersion
  )
}

private func score(_ record: OperonMemoryRecord, terms: [String]) -> Int {
  let text = "\(record.kind.rawValue) \(record.subject ?? "") \(record.content)".lowercased()
  return terms.reduce(into: 0) { value, term in
    value += text.components(separatedBy: term).count - 1
  }
}

File: sdk/python/src/operon/memory.py
"""Typed, application-authorized durable memory for Operon hosts.

Records are created by application code. This module deliberately has no model
write path: an LLM may consume retrieved records as historical context, but it
does not gain authority to create, update, or delete them.
"""

from __future__ import annotations

import json
import re
import sqlite3
import uuid
from dataclasses import dataclass, replace
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Protocol


class MemoryKind(str, Enum):
    FACT = "fact"
    PREFERENCE = "preference"
    DECISION = "decision"
    EPISODE = "episode"


class MemoryAuthority(str, Enum):
    APPLICATION_VERIFIED = "application_verified"
    USER_CONFIRMED = "user_confirmed"
    USER_STATED = "user_stated"
    MODEL_INFERRED = "model_inferred"
    IMPORTED_UNTRUSTED = "imported_untrusted"


class MemorySensitivity(str, Enum):
    PRIVATE = "private"
    INTERNAL = "internal"
    PUBLIC = "public"


class MemoryStatus(str, Enum):
    ACTIVE = "active"
    SUPERSEDED = "superseded"
    TOMBSTONED = "tombstoned"


@dataclass(frozen=True, slots=True)
class MemoryScope:
    """The authorization boundary applied before relevance ranking."""

    namespace: str
    subject: str | None = None
    allowed_sensitivities: tuple[MemorySensitivity, ...] = (
        MemorySensitivity.PRIVATE,
        MemorySensitivity.INTERNAL,
    )

    def __post_init__(self) -> None:
        if not self.namespace.strip():
            raise ValueError("memory namespace cannot be empty")
        if not self.allowed_sensitivities:
            raise ValueError("memory scope needs at least one allowed sensitivity")


@dataclass(frozen=True, slots=True)
class MemoryRecord:
    id: str
    namespace: str
    subject: str | None
    kind: MemoryKind
    content: str
    authority: MemoryAuthority
    sensitivity: MemorySensitivity = MemorySensitivity.PRIVATE
    confidence: float | None = None
    source_ids: tuple[str, ...] = ()
    occurred_at: str | None = None
    observed_at: str | None = None
    valid_from: str | None = None
    valid_until: str | None = None
    supersedes: str | None = None
    status: MemoryStatus = MemoryStatus.ACTIVE
    created_by: str = "application"
    schema_version: int = 1

    @classmethod
    def create(
        cls,
        *,
        namespace: str,
        kind: MemoryKind,
        content: str,
        authority: MemoryAuthority,
        subject: str | None = None,
        **kwargs: object,
    ) -> MemoryRecord:
        return cls(
            id=str(uuid.uuid4()),
            namespace=namespace,
            subject=subject,
            kind=kind,
            content=content,
            authority=authority,
            **kwargs,
        )


@dataclass(frozen=True, slots=True)
class MemoryContext:
    scope: MemoryScope
    records: tuple[MemoryRecord, ...]
    text: str
    omitted_record_count: int


class MemoryStore(Protocol):
    def put(self, record: MemoryRecord) -> MemoryRecord: ...

    def search(
        self, query: str, scope: MemoryScope, limit: int
    ) -> tuple[MemoryRecord, ...]: ...

    def context(
        self, query: str, scope: MemoryScope, limit: int, maximum_characters: int
    ) -> MemoryContext: ...

    def tombstone(self, memory_id: str) -> bool: ...

    def export(self, scope: MemoryScope) -> dict[str, object]: ...

    def delete_namespace(self, namespace: str) -> int: ...


class SQLiteMemoryStore:
    """A local SQLite/FTS5 store for explicit, typed durable memory records."""

    def __init__(self, path: str | Path) -> None:
        self.path = str(path)
        if self.path == ":memory:":
            raise ValueError("SQLiteMemoryStore requires a file path, not :memory:")
        Path(self.path).expanduser().parent.mkdir(parents=True, exist_ok=True)
        self._initialize()

    def put(self, record: MemoryRecord) -> MemoryRecord:
        record = _validated_record(record)
        with self._connect() as connection:
            connection.execute("BEGIN IMMEDIATE")
            if record.supersedes:
                connection.execute(
                    """
                    UPDATE operon_memory
                    SET status = ?
                    WHERE id = ? AND namespace = ? AND status = ?
                    """,
                    (
                        MemoryStatus.SUPERSEDED.value,
                        record.supersedes,
                        record.namespace,
                        MemoryStatus.ACTIVE.value,
                    ),
                )
                connection.execute(
                    "DELETE FROM operon_memory_fts WHERE memory_id = ?",
                    (record.supersedes,),
                )
            connection.execute(
                """
                INSERT INTO operon_memory(
                    id, namespace, subject, kind, content, authority, sensitivity,
                    confidence, source_ids, occurred_at, observed_at, valid_from,
                    valid_until, supersedes, status, created_by, schema_version
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                _record_row(record),
            )
            connection.execute(
                "INSERT INTO operon_memory_fts(memory_id, search_text) VALUES (?, ?)",
                (record.id, _search_text(record)),
            )
        return record

    def search(
        self, query: str, scope: MemoryScope, limit: int
    ) -> tuple[MemoryRecord, ...]:
        if limit < 1:
            raise ValueError("memory search limit must be positive")
        match = _fts_query(query)
        if not match:
            return ()
        filters = [
            "m.namespace = ?",
            "m.status = ?",
            "(m.valid_from IS NULL OR m.valid_from <= ?)",
            "(m.valid_until IS NULL OR m.valid_until > ?)",
        ]
        values: list[object] = [
            scope.namespace,
            MemoryStatus.ACTIVE.value,
            _now(),
            _now(),
        ]
        if scope.subject is not None:
            filters.append("m.subject = ?")
            values.append(scope.subject)
        placeholders = ", ".join("?" for _ in scope.allowed_sensitivities)
        filters.append(f"m.sensitivity IN ({placeholders})")
        values.extend(item.value for item in scope.allowed_sensitivities)
        values.extend([match, limit])
        sql = f"""
            SELECT m.id, m.namespace, m.subject, m.kind, m.content, m.authority,
                   m.sensitivity, m.confidence, m.source_ids, m.occurred_at,
                   m.observed_at, m.valid_from, m.valid_until, m.supersedes,
                   m.status, m.created_by, m.schema_version
            FROM operon_memory_fts
            JOIN operon_memory AS m ON m.id = operon_memory_fts.memory_id
            WHERE {' AND '.join(filters)}
              AND operon_memory_fts.search_text MATCH ?
            ORDER BY bm25(operon_memory_fts), m.observed_at DESC
            LIMIT ?
        """
        with self._connect() as connection:
            rows = connection.execute(sql, values).fetchall()
        return tuple(_row_record(row) for row in rows)

    def context(
        self, query: str, scope: MemoryScope, limit: int, maximum_characters: int
    ) -> MemoryContext:
        if maximum_characters < 1:
            raise ValueError("maximum_characters must be positive")
        records = self.search(query, scope, limit)
        sections: list[str] = []
        remaining = maximum_characters
        for record in records:
            text = _render(record)
            if len(text) > remaining:
                break
            sections.append(text)
            remaining -= len(text) + 2
        included = records[: len(sections)]
        return MemoryContext(
            scope=scope,
            records=included,
            text="\n\n".join(sections),
            omitted_record_count=len(records) - len(included),
        )

    def tombstone(self, memory_id: str) -> bool:
        memory_id = memory_id.strip()
        if not memory_id:
            raise ValueError("memory_id cannot be empty")
        with self._connect() as connection:
            connection.execute("BEGIN IMMEDIATE")
            updated = connection.execute(
                "UPDATE operon_memory SET status = ? WHERE id = ? AND status != ?",
                (MemoryStatus.TOMBSTONED.value, memory_id, MemoryStatus.TOMBSTONED.value),
            ).rowcount
            connection.execute(
                "DELETE FROM operon_memory_fts WHERE memory_id = ?", (memory_id,)
            )
        return bool(updated)

    def export(self, scope: MemoryScope) -> dict[str, object]:
        filters = ["namespace = ?"]
        values: list[object] = [scope.namespace]
        if scope.subject is not None:
            filters.append("subject = ?")
            values.append(scope.subject)
        with self._connect() as connection:
            rows = connection.execute(
                "SELECT id, namespace, subject, kind, content, authority, sensitivity, "
                "confidence, source_ids, occurred_at, observed_at, valid_from, valid_until, "
                "supersedes, status, created_by, schema_version FROM operon_memory WHERE "
                + " AND ".join(filters)
                + " ORDER BY observed_at ASC, id ASC",
                values,
            ).fetchall()
        return {
            "schema_version": 1,
            "scope": {"namespace": scope.namespace, "subject": scope.subject},
            "records": [_record_export(_row_record(row)) for row in rows],
        }

    def delete_namespace(self, namespace: str) -> int:
        namespace = namespace.strip()
        if not namespace:
            raise ValueError("memory namespace cannot be empty")
        with self._connect() as connection:
            connection.execute("BEGIN IMMEDIATE")
            ids = connection.execute(
                "SELECT id FROM operon_memory WHERE namespace = ?", (namespace,)
            ).fetchall()
            connection.executemany(
                "DELETE FROM operon_memory_fts WHERE memory_id = ?", ids
            )
            deleted = connection.execute(
                "DELETE FROM operon_memory WHERE namespace = ?", (namespace,)
            ).rowcount
        return deleted

    def _initialize(self) -> None:
        with self._connect() as connection:
            connection.executescript(
                """
                CREATE TABLE IF NOT EXISTS operon_memory (
                    id TEXT PRIMARY KEY,
                    namespace TEXT NOT NULL,
                    subject TEXT,
                    kind TEXT NOT NULL,
                    content TEXT NOT NULL,
                    authority TEXT NOT NULL,
                    sensitivity TEXT NOT NULL,
                    confidence REAL,
                    source_ids TEXT NOT NULL,
                    occurred_at TEXT,
                    observed_at TEXT NOT NULL,
                    valid_from TEXT,
                    valid_until TEXT,
                    supersedes TEXT,
                    status TEXT NOT NULL,
                    created_by TEXT NOT NULL,
                    schema_version INTEGER NOT NULL
                );
                CREATE INDEX IF NOT EXISTS operon_memory_by_scope
                    ON operon_memory(namespace, subject, status, observed_at);
                CREATE VIRTUAL TABLE IF NOT EXISTS operon_memory_fts USING fts5(
                    memory_id UNINDEXED,
                    search_text
                );
                """
            )

    def _connect(self) -> sqlite3.Connection:
        return sqlite3.connect(self.path)


def _validated_record(record: MemoryRecord) -> MemoryRecord:
    if not record.id.strip():
        raise ValueError("memory id cannot be empty")
    if not record.namespace.strip():
        raise ValueError("memory namespace cannot be empty")
    if not record.content.strip():
        raise ValueError("memory content cannot be empty")
    if record.confidence is not None and not 0 <= record.confidence <= 1:
        raise ValueError("memory confidence must be between 0 and 1")
    if record.valid_from and record.valid_until and record.valid_from >= record.valid_until:
        raise ValueError("memory valid_until must be later than valid_from")
    return replace(record, observed_at=record.observed_at or _now())


def _record_row(record: MemoryRecord) -> tuple[object, ...]:
    return (
        record.id,
        record.namespace,
        record.subject,
        record.kind.value,
        record.content,
        record.authority.value,
        record.sensitivity.value,
        record.confidence,
        json.dumps(record.source_ids),
        record.occurred_at,
        record.observed_at,
        record.valid_from,
        record.valid_until,
        record.supersedes,
        record.status.value,
        record.created_by,
        record.schema_version,
    )


def _row_record(row: tuple[object, ...]) -> MemoryRecord:
    return MemoryRecord(
        id=str(row[0]),
        namespace=str(row[1]),
        subject=row[2] if isinstance(row[2], str) else None,
        kind=MemoryKind(str(row[3])),
        content=str(row[4]),
        authority=MemoryAuthority(str(row[5])),
        sensitivity=MemorySensitivity(str(row[6])),
        confidence=float(row[7]) if row[7] is not None else None,
        source_ids=tuple(json.loads(str(row[8]))),
        occurred_at=row[9] if isinstance(row[9], str) else None,
        observed_at=row[10] if isinstance(row[10], str) else None,
        valid_from=row[11] if isinstance(row[11], str) else None,
        valid_until=row[12] if isinstance(row[12], str) else None,
        supersedes=row[13] if isinstance(row[13], str) else None,
        status=MemoryStatus(str(row[14])),
        created_by=str(row[15]),
        schema_version=int(row[16]),
    )


def _record_export(record: MemoryRecord) -> dict[str, object]:
    return {
        "id": record.id,
        "namespace": record.namespace,
        "subject": record.subject,
        "kind": record.kind.value,
        "content": record.content,
        "authority": record.authority.value,
        "sensitivity": record.sensitivity.value,
        "confidence": record.confidence,
        "source_ids": list(record.source_ids),
        "occurred_at": record.occurred_at,
        "observed_at": record.observed_at,
        "valid_from": record.valid_from,
        "valid_until": record.valid_until,
        "supersedes": record.supersedes,
        "status": record.status.value,
        "created_by": record.created_by,
        "schema_version": record.schema_version,
    }


def _search_text(record: MemoryRecord) -> str:
    return " ".join(
        part for part in (record.kind.value, record.content, record.subject or "") if part
    )


def _fts_query(query: str) -> str:
    terms = re.findall(r"[\w]+", query.lower(), flags=re.UNICODE)
    return " OR ".join(f'"{term}"' for term in terms)


def _render(record: MemoryRecord) -> str:
    subject = f" subject={record.subject}" if record.subject else ""
    time = f" observed_at={record.observed_at}" if record.observed_at else ""
    return (
        f"[M:{record.id}] kind={record.kind.value} authority={record.authority.value}"
        f" sensitivity={record.sensitivity.value}{subject}{time}\n{record.content}"
    )


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()
